From bd38fa9829cd8500a08a15835017d0adc254c4ed Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Tue, 6 Dec 2011 20:44:13 +0100
Subject: [PATCH 051/124] libmythtv: Add a mutex to the ChannelScanSM class

This class is accessed by two threads, the signal monitor and the
thread that drives the scan.

A segv can occur if one thread deletes the currentInfo member
while it is being used by the other.

Signed-off-by: Lawrence Rust <lvr@softsystem.co.uk>
---
 .../libs/libmythtv/channelscan/channelscan_sm.cpp  |    3 +++
 mythtv/libs/libmythtv/channelscan/channelscan_sm.h |    2 ++
 2 files changed, 5 insertions(+)

diff --git a/mythtv/libs/libmythtv/channelscan/channelscan_sm.cpp b/mythtv/libs/libmythtv/channelscan/channelscan_sm.cpp
index 2328bf3..d37e5a9 100644
--- a/mythtv/libs/libmythtv/channelscan/channelscan_sm.cpp
+++ b/mythtv/libs/libmythtv/channelscan/channelscan_sm.cpp
@@ -36,6 +36,7 @@ using namespace std;
 
 // Qt includes
 #include <QObject>
+#include <QMutexLocker>
 
 // MythTV includes - General
 #include "channelscan_sm.h"
@@ -724,6 +725,8 @@ void ChannelScanSM::UpdateScanTransports(const NetworkInformationTable *nit)
 
 bool ChannelScanSM::UpdateChannelInfo(bool wait_until_complete)
 {
+    QMutexLocker locker(&m_mutex);
+
     if (current == scanTransports.end())
         return true;
 
diff --git a/mythtv/libs/libmythtv/channelscan/channelscan_sm.h b/mythtv/libs/libmythtv/channelscan/channelscan_sm.h
index 459cf02..5edb2b0 100644
--- a/mythtv/libs/libmythtv/channelscan/channelscan_sm.h
+++ b/mythtv/libs/libmythtv/channelscan/channelscan_sm.h
@@ -37,6 +37,7 @@
 #include <QPair>
 #include <QMap>
 #include <QSet>
+#include <QMutex>
 
 // MythTV includes
 #include "frequencytables.h"
@@ -248,6 +249,7 @@ class ChannelScanSM : public MPEGStreamListener,
 
     /// Scanner thread, runs ChannelScanSM::run()
     MThread          *scannerThread;
+    QMutex           m_mutex;
 };
 
 inline void ChannelScanSM::UpdateScanPercentCompleted(void)
-- 
1.7.9.5

